compiler/optimizer/VPHandlers.cpp

```diff
@@ -1151,31 +1151,86 @@ TR::Node *constrainBCDAggrLoad(OMR::ValuePropagation *vp, TR::Node *node)
-      if (index->getOpCode().isLoadConst() && array->getOpCode().isIndirect())
+      if (index->getOpCode().isLoadConst())
          {
          bool isGlobal;
-         TR::Node* base = array->getFirstChild();
-         TR::VPConstraint *baseVPConstraint = vp->getConstraint(base, isGlobal);
-         if (baseVPConstraint && baseVPConstraint->isConstString())
+         if (dataType == TR::Int16 && array->getOpCode().isIndirect())
             {
-            TR::VPConstString *constString = baseVPConstraint->getClassType()->asConstString();
+            TR::Node* base = array->getFirstChild();
+            TR::VPConstraint *baseVPConstraint = vp->getConstraint(base, isGlobal);
+            if (baseVPConstraint && baseVPConstraint->isConstString())
+               {
+               TR::VPConstString *constString = baseVPConstraint->getClassType()->asConstString();
 
-            uint16_t ch = constString->charAt(((TR::Compiler->target.is64Bit() ? index->getLongIntLow() : index->getInt())
-                                               - TR::Compiler->om.contiguousArrayHeaderSizeInBytes()) / 2, vp->comp());
-            if (ch != 0)
+               uint16_t ch = constString->charAt(((TR::Compiler->target.is64Bit() ? index->getLongIntLow() : index->getInt())
+                                                  - TR::Compiler->om.contiguousArrayHeaderSizeInBytes()) / 2, vp->comp());
+               if (ch != 0)
+                  {
+                  vp->replaceByConstant(node, TR::VPShortConst::create(vp, ch), isGlobal);
+                  return node;
+                  }
+               }
+            }
+         TR::VPConstraint *arrayVPConstraint = vp->getConstraint(array, isGlobal);
+         if (arrayVPConstraint)
+            {
+#ifdef J9_PROJECT_SPECIFIC
+            TR::KnownObjectTable *knot = vp->comp()->getKnownObjectTable();
+            TR::VPKnownObject *kobj = arrayVPConstraint->getKnownObject();
+            if (knot && kobj)
                {
-               vp->replaceByConstant(node, TR::VPIntConst::create(vp, ch), true);
-               return node;
+               TR::KnownObjectTable::Index idx = kobj->getIndex();
+               if (kobj->isArrayWithConstantElements(vp->comp()) && kobj->isPrimitiveArray(vp->comp()))
+                  {
+                  TR::VMAccessCriticalSection constrainAnyIntLoadCriticalSection(vp->comp(),
+                     TR::VMAccessCriticalSection::tryToAcquireVMAccess);
+                  if (constrainAnyIntLoadCriticalSection.hasVMAccess())
+                     {
+                     uintptrj_t arrayObj = knot->getPointer(idx);
+                     uintptrj_t offset = TR::Compiler->target.is64Bit()
+                                         ? (uintptrj_t)index->getUnsignedLongInt() : (uintptrj_t)index->getUnsignedInt();
+                     uintptrj_t lengthInElements = TR::Compiler->om.getArrayLengthInElements(vp->comp(), arrayObj);
```

Comment:
What's the rationale for the getUnsigned* usages here ?

Reply:
Just to make it obvious that there are no sign-extensions happening. By calling `getUnsigned*`, we reinterpret the constant as unsigned before casting to `uintptrj_t`, which may be a different size. Using `getLongInt()` or `getInt()` also wouldn't sign-extend, but only because at runtime we'll run the branch where `uintptrj_t` is the same size as the constant.

Reply:
Now that I've described the index calculation for `charAt()` above, I'll add that those are the kinds of shenanigans I wanted to make clear were not going on here.